network/p2p/sender.go

```diff
@@ -0,0 +1,210 @@
+
+import (
+	"context"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+var (
+	_ AppSender = (*Sender)(nil)
+	_ AppSender = (*FakeSender)(nil)
+)
+
+type AppRequestSender interface {
+	// SendAppRequest sends an AppRequest message to a peer
+	SendAppRequest(ctx context.Context, nodeID ids.NodeID, requestID uint32, bytes []byte) error
+}
+
+type AppResponseSender interface {
+	// SendAppRequest sends an AppResponse message to a peer
+	SendAppResponse(ctx context.Context, nodeID ids.NodeID, requestID uint32, bytes []byte) error
+}
+
+type AppGossipSender interface {
+	// SendAppGossip sends an AppGossip message to an arbitrary peer
+	SendAppGossip(ctx context.Context, bytes []byte) error
+	// SendAppGossipSpecific sends an AppGossip message to a given peer
+	SendAppGossipSpecific(ctx context.Context, nodeID ids.NodeID, bytes []byte) error
+}
+
+type CrossChainAppRequestSender interface {
+	// SendCrossChainAppRequest sends a CrossChainAppRequest  message to a
+	// local chain
+	SendCrossChainAppRequest(ctx context.Context, chainID ids.ID, requestID uint32, bytes []byte) error
+}
+
+type CrossChainAppResponseSender interface {
+	// SendCrossChainAppResponse sends a CrossChainAppResponse message to a
+	// local chain
+	SendCrossChainAppResponse(ctx context.Context, chainID ids.ID, requestID uint32, bytes []byte) error
+}
+
+// AppSender sends p2p messages
+type AppSender interface {
+	AppRequestSender
+	AppResponseSender
+	AppGossipSender
+	CrossChainAppRequestSender
+	CrossChainAppResponseSender
+}
+
+// NewSender returns an instance of Sender
+func NewSender(sender common.AppSender) *Sender {
+	return &Sender{
+		sender: sender,
+	}
+}
+
+// Sender sends messages over the p2p network
+type Sender struct {
+	sender common.AppSender
+}
+
+func (s Sender) SendAppRequest(ctx context.Context, nodeID ids.NodeID, requestID uint32, bytes []byte) error {
+	return s.sender.SendAppRequest(ctx, set.Of(nodeID), requestID, bytes)
+}
+
+func (s Sender) SendAppResponse(ctx context.Context, nodeID ids.NodeID, requestID uint32, bytes []byte) error {
+	return s.sender.SendAppResponse(ctx, nodeID, requestID, bytes)
+}
+
+func (s Sender) SendAppGossip(ctx context.Context, bytes []byte) error {
+	return s.sender.SendAppGossip(ctx, bytes)
+}
+
+func (s Sender) SendAppGossipSpecific(ctx context.Context, nodeID ids.NodeID, bytes []byte) error {
+	return s.sender.SendAppGossipSpecific(ctx, set.Of(nodeID), bytes)
+}
+
+func (s Sender) SendCrossChainAppRequest(ctx context.Context, chainID ids.ID, requestID uint32, bytes []byte) error {
+	return s.sender.SendCrossChainAppRequest(ctx, chainID, requestID, bytes)
+}
+
+func (s Sender) SendCrossChainAppResponse(ctx context.Context, chainID ids.ID, requestID uint32, bytes []byte) error {
+	return s.sender.SendCrossChainAppResponse(ctx, chainID, requestID, bytes)
+}
+
+// FakeSender is used for testing
+type FakeSender struct {
```

Comment:
Do we need both `FakeSender` and `MockSender`? I think we could get the same functionality that `FakeSender` has (i.e. piping messages to a channel) with `MockSender` by defining the mock function definitions that way.

Reply:
I agree we should ideally pick one way of testing and stick with it (I'd vote for no mocks anymore).
But it's a more general discussion and we may defer the cleanup to a different PR (even multiple ones) once we agree what testing strategy should we pick?

Reply:
I prefer to use this channel pattern wherever possible to avoid the goroutine stuff... I think I'm only using MockSender in spots where we do an assertion against the nodeID